Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,218 @@
+
+### Local Feature Switches boilerplate
+
+It takes a lot of boilerplate to define a local feature switch - we need to change the code in several places (in one file though), need to manually add an entry to the `Root.plist` and make sure that it's in the right place and in the write file. We need to make sure we move it to the different section (for remote feature toggles) if we move the flag to the Firebase, otherwise we won't be able to override it in the Settings app and it will always have a default value if we override any other flag.
+
+
+### Debugging
+
+It's not clear what feature switches affect what screens. It's somtimes frustrating to go through a complex flow and then realise that the right feature switch was not turned on/off in the Settings app or local overrides were active and to go through the flow again (as we don't observe changes in the feature switches at runtime and require application to restart).
+
+
+## Proposed solution
+
+To address these issues we suggest the following changes in our feature switches architecture:
+
+1. Feature toggles and static configurations related to the particular feature should be defined in this feature module instead of shared module or in the app. This follows interface segregation principle and goes along with improving modularisation - one won't need to change common module with feature switches and recompile all of it (and potentially everything else) when they only need to recompile module that uses this feature switch. This way we also configure our frameworks the same way as we configure any other 3rd party framework and can use the similar infrastrauctore for that (i.e. dedicated instances responsibule for configuring each feature rather than doing that all in the app delegate).
+
+2. Replace `AppConfigurationProtocol` and all related protocols with structs. This will decrease amount of boilerplate we have to write when adding new application configuration. Application configuration will be only concerned with application level configurations which are not directly related to any feature module.
+
+3. All feature switches access points should be moved to the `Current` and consequentially should be only used by builders - view models, flow controllers and renderers should have them injected through their constructors.
+
+4. Root.plist should be generated based on the code that declares feature switches.
+
+## Detailed design
+
+### Module specific feature switches
+
+Each type of feature switch will be implemented with it's own type to make its semantics clear. We will need `LocalFeatureSwitch` and `RemoteFeatureSwitch` types. When later we introduce a new produt configurator service we defined new type `ProductFeatureSwitch`. They all will have a common `value` property and each provider will have a `get(valueForKey: String)` method and should store their properties in a dictionary.
+
+```swift
+// In BabylonDependencies
+
+public struct RemoteFeatureSwitch<T> {
```

Comment:
Those `*FeatureSwitch` types don't seem much different. Would it make sense to have a single type with extra generic parameter instead? I think this keeps readability and reduces boilerplate a little bit.

```swift
struct FeatureSwitch<T, Source> {
    let name: String
    let key: String
}

enum Remote {}
enum Local {}

extension FeatureSwitch where Source == Remote {
    var value: Bool {
        return Current.abTestingService.get(valueForKey: self.key)
    }
}

extension FeatureSwitch where Source == Local {
    var value: Bool {
        return Current.userDefaults.get(valueForKey: self.key)
    }
}

let remoteFeatureSwitch = FeatureSwitch<Bool, Remote>(name: "New feature", key: "is_new_feature_enabled")
let localFeatureSwitch = FeatureSwitch<Bool, Local>(name: "Another feature", key: "is_another_feature_enabled")
```

Reply:
The "phantom" type would work as well for that purpose, but IMO declaration becomes a little bit less nice. And I'm not sure how nice it will play with property wrappers (which we can easily convert to as they have exactly the same signature), I think they allow only one generic. So then it can be `@RemoteFeatureSwitch` or `@LocalFeatureSwitch`. I don't think it's possible to do `@FeatureSwitch<Local/Remote>`.

Reply:
Though we can use generic type alias for this purpose:

```
typealias RemoteFeatureSwitch<T> = FeatureSwitch<T, Remote>
typealias LocalFeatureSwitch<T> = FeatureSwitch<T, Local>
```

Not sure still if it will work with property wrappers, but they are not a requirement of this redesign, so we can as well go with one generic implementation as you suggest.

Reply:
FWIW Property wrappers support multiple generics, but the problem is inference. With one single generic parameter `T` that is used by the `wrappedValue: T`, the compiler can infer `T` from the type of the variable. But if one of the generic parameter is a phantom type, not used by the wrapped value, and not inferable by a parameter passed to the PW attribute, then you have to specify the full list of generic parameters to the PW attribute.

So we could totally do this:

```swift
@FeatureSwitch<Bool, FeatureSwitchLocal>(name: "Shiny Feature", key: "shiny_feature")
var shinyFeatureEnabled: Bool
```

But it would be a shame to have to specify the full generic parameters instead of relying on inference like we could do without the phantom type, letting the compiler infer the `Bool` here:

```swift
@LocalFeatureSwitch(name: "Shiny Feature", key: "shiny_feature")
var shinyFeatureEnabled: Bool
```

So not so sure a phantom type is worth the trouble anyway indeed. And having separate types could also allow us to expose other methods on the `projectedValue` if we wanted to, like `$shinyFeatureEnabled.toggle()` or `$shinyFeatureEnabled.someAdditionalMetaDataThatWouldOnlyMakeSenseForLocalFS`